[ParseableInterface] Make the module caches relocatable and respect -track-system-dependencies

nathawes opened the pull request:
1. Honor `-track-system-dependencies` when building parseable interfaces on demand
    - FIXME: won't be tracked when the top-level invocation isn't tracking dependencies at all
2. Honor `-track-system-dependencies` with `-build-module-from-parseable-interface`
3. Serialize dependencies in the SDK as SDK-relative paths
    - These are expanded to full paths by appending the SDK path when writing out forwarding modules and when checking dependencies are up-to-date.
4. Stop serializing `.swiftmodule` dependencies that are from the regular and prebuilt module caches
    - Note: `.swiftmodule`s in the regular module cache still get reported to the dependency tracker. Is this what we want?

Resolves rdar://problem/49292914

jrose-apple committed:
[ParseableInterface] Honor -track-system-dependencies when caching

One interesting thing here is that the decision of whether or not to
track system dependencies ends up going into the cache key for a
swiftmodule built from a parseable interface, because it affects that
module's up-to-date check. If we didn't include
-track-system-dependencies in the cache key, we could end up tracking
system dependencies for some modules but not others in the same build.

There's a bit of a bug here where they're /not/ honored if the
top-level invocation isn't tracking /any/ dependencies, but given
how uncommon this flag is in practice that's probably okay for now.

Still TODO: honor this for -build-swiftmodule-from-parseable-interface
as well. That's currently not tracking dependencies at all and it
probably should.

Commit:
[ParseableInterface] Don't serialize swiftmodule dependencies in the module cache or prebuilt module cache

*Their* dependencies are already being serialized out, so this shouldn't affect
up-to-date-checking except by alowing the regular and prebuilt module caches to
be relocated without invalidating their contents. In the case of the prebuilt
module cache, this gets us closer to supporting relocation across machines.

Commit:
[ParseableInterface] Respect -track-system-dependencies with -build-module-from-parseable-interface

Updates the subinvocation that builds the parseable interface to respect the
-track-system-dependencies flag of the top-level invocation if present, by
including system dependencies in the produced .swiftmodule.

Commit:
[ParseableInterface] Distinguish SDK and non-SDK dependencies

This allows the SDK to be relocated without automatically resulting in a
rebuild.

Based on an old patch from Jordan Rose.

nathawes commented:
@swift-ci please test

jrose-apple commented on test/ParseableInterface/ModuleCache/SDKDependencies.swiftinterface:
```diff
@@ -0,0 +1,123 @@
+//
+// Check they are *not* forwarding modules
+// RUN: head -n 1 %t/MCP/SdkLib-*.swiftmodule | not grep -e '---'
+// RUN: head -n 1 %t/MCP/ExportedLib-*.swiftmodule | not grep -e '---'
```
Harlan wrote a test for this, `%S/Inputs/check-is-fowarding-module.py`. It does basically the same thing but is a little more semantic.

nathawes replied:
Thanks! Updated.

jrose-apple commented on lib/Frontend/ParseableInterfaceModuleLoader.cpp:
```diff
@@ -356,48 +356,36 @@ class swift::ParseableInterfaceBuilder {
+  /// either the module cache or prebuilt module cache
+  bool isCachedModule(StringRef DepName) const {
+    if (moduleCachePath.empty() && prebuiltCachePath.empty())
+        return false;
```
Double indent snuck in here.

jrose-apple commented on lib/Frontend/ParseableInterfaceModuleLoader.cpp:
```diff
@@ -356,48 +356,36 @@ class swift::ParseableInterfaceBuilder {
+    auto Ty = file_types::lookupTypeForExtension(Ext);
+    return Ty == file_types::TY_SwiftModuleFile &&
+        ((!moduleCachePath.empty() && DepName.startswith(moduleCachePath)) ||
+         (!prebuiltCachePath.empty() && DepName.startswith(prebuiltCachePath)));
```
Nitpick: I think this might be clearer as multiple `if` statements, rather than trying to make one big expression.

jrose-apple commented on test/ParseableInterface/ModuleCache/SDKDependencies.swift:
```diff
@@ -74,11 +71,9 @@
 // NEW-EXLIB-DAG: /my-new-sdk/usr/include/module.modulemap
 // NEW-EXLIB-DAG: /my-new-sdk/usr/include/SomeCModule.h
 // NEW-EXLIB-DAG: /my-new-sdk/ExportedLib.swiftinterface
-// NEW-SDKLIB-DAG: /prebuilt-cache/SdkLib.swiftmodule
```
Can you do some negative checks to make sure that these entries are *not* there?

nathawes replied:
Added the negative checks and also updated it to check the dependencies in the serialized modules too.

jrose-apple reviewed (commented):
I did a quick read and it looks pretty good. @harlanhaskins should probably take a look too, and I'll get back to it a second time later today.

jrose-apple commented:
> Note: .swiftmodules in the regular module cache still get reported to the dependency tracker. Is this what we want?

I've come to a conclusion on this: if we compile the prebuilt modules out of order, they will load swiftmodules from the regular module cache, and it would be bad to mark those as dependencies.

harlanhaskins commented on lib/Frontend/ParseableInterfaceModuleLoader.cpp:
```diff
@@ -702,10 +742,17 @@ class ParseableInterfaceModuleLoaderImpl {
   bool dependenciesAreUpToDate(ArrayRef<FileDependency> deps) {
+    SmallString<128> SDKRelativeBuffer;
     for (auto &in : deps) {
+      StringRef fullPath = in.getPath();
```
Nitpick: This logic is repeated in `writeForwardingModule` -- does it make sense to move it into FileDependency/a helper?

nathawes replied:
Moved it out into a helper

harlanhaskins commented on test/ParseableInterface/ModuleCache/SDKDependencies.swift:
```diff
@@ -0,0 +1,118 @@
+// RUN: cat %t/MCP/SdkLib-*.swiftmodule | %FileCheck %s -check-prefixes=EXLIB,SDKLIB
+//
+// EXLIB: dependencies:
+// EXLIB-DAG: /my-sdk/usr/include/module.modulemap
```
Can you add a `NOT` line to make sure we don't record any `.swiftmodule` files from the cache?

harlanhaskins reviewed (approved):
Small comment, otherwise this looks great!

Commit:
[ParseableInterface][test] Update SDKDependencies.swift test to explicitly check serialized dependencies

Also use the existing check-is-forwarding-module.py script to check for
forwarding modules, rather than doing it manually.

nathawes commented:
> > Note: .swiftmodules in the regular module cache still get reported to the dependency tracker. Is this what we want?
> 
> I've come to a conclusion on this: if we compile the prebuilt modules out of order, they will load swiftmodules from the regular module cache, and it would be bad to mark those as dependencies.

Good point! I've updated the SDKDependencies.swift test to building the prebuilt modules in both orders and check the serialized modules don't have any FILE_DEPENDENCY entries containing the module cache path in either case.

jrose-apple commented on lib/Frontend/ParseableInterfaceModuleLoader.cpp:
```diff
@@ -356,48 +356,39 @@ class swift::ParseableInterfaceBuilder {
+  /// Determines if the dependency with the provided path is a swiftmodule in
+  /// either the module cache or prebuilt module cache
+  bool isCachedModule(StringRef DepName) const {
+    if (moduleCachePath.empty() && prebuiltCachePath.empty())
```
Nitpick: I think in practice `moduleCachePath` will never be empty. I wouldn't remove the check from below just in case but we probably don't need the early exit here.

jrose-apple commented on test/ParseableInterface/ModuleCache/SDKDependencies.swift:
```diff
@@ -20,8 +38,12 @@
-// RUN: head -n 1 %t/MCP/SdkLib-*.swiftmodule | not grep -e '---'
-// RUN: head -n 1 %t/MCP/ExportedLib-*.swiftmodule | not grep -e '---'
+// RUN: not python %S/Inputs/check-is-forwarding-module.py %t/MCP/SdkLib-*.swiftmodule
+// RUN: not python %S/Inputs/check-is-forwarding-module.py %t/MCP/ExportedLib-*.swiftmodule
```
Nitpick: `%{python}` ensures that we use the same Python being used to run lit.py, which can be important on systems with Python 3 installed.

jrose-apple commented on test/ParseableInterface/ModuleCache/SDKDependencies.swift:
```diff
@@ -49,23 +71,33 @@
+// NOCACHE: dependencies:
+// NOCACHE: /prebuilt-cache/[[LIB_NAME]].swiftmodule
+// NOCACHE-NOT: /prebuilt-cache/
+// NOCACHE-NOT: /MCP/
```
The way negative lit checks work is that they can't appear between two positive checks, or between the last positive check and the end of the file. So this doesn't check that there aren't more dependencies *before* the one you've named.

jrose-apple reviewed (commented):
Okay, went through it again and I think it's good! Minor comments only.

nathawes commented:
@swift-ci please test OS X Platform

nathawes commented:
@swift-ci please test OS X Platform

nathawes commented:
@swift-ci please smoke test OS X Platform

Commit:
[ParseableInterface] Don't serialize resource directory deps and stop adding cached modules to the dependency tracker

This patch modifies ParseableInterfaceBuilder::CollectDepsForSerialization to
avoid serializing dependencies from the runtime resource path into the
swiftmodules generated from .swiftinterface files. This means the module cache
should now be relocatable across machines.

It also modifies ParseableInterfaceModuleLoader to never add any dependencies
from the module cache and prebuilt cache to the dependency tracker (in addition
to the existing behaviour of not serializing them in the generated
swiftmodules). As a result, CollectDepsForSerialization no longer checks if the
dependencies it is given come from the cache as they are provided by the
dependency tracker. It now asserts that's the case instead.

nathawes commented:
@swift-ci please test

nathawes commented on lib/Frontend/ParseableInterfaceModuleLoader.cpp:
```diff
@@ -399,44 +430,14 @@ class swift::ParseableInterfaceBuilder {
-      // If Dep is itself a .swiftmodule in the cache dir, pull out its deps
-      // and include them in our own, so we have a single-file view of
-      // transitive deps: removes redundancies, and avoids opening and reading
-      // multiple swiftmodules during future loads.
```
It seems like pulling these in was redundant? I think everything still works because whenever we load/depend on something from the cache we end up adding its serialized dependencies to the dependency tracker when validating that it's up to date, so they show up as one of the initial dependencies here and get serialized.

jrose-apple replied:
That doesn't sound right. What if it's out of date and then we rebuild it and discover it doesn't depend on some of those files anymore?

nathawes replied:
`ParseableInterfaceModuleLoaderImpl::dependenciesAreUpToDate()` is called by both `forwardingModuleIsUpToDate()` and `swiftModuleIsUpDate()` and adds the dependencies to the tracker regardless of whether they are out of date or not, so looks like that's a bug. I'll add a test for it and update the code to only add them to the tracker if they're all up to date.

In the case that they're not up to date though, we then go down the path of generating the swiftmodule from the interface again, and that adds all the `SubInstance`'s dependencies to the tracker (in `collectDepsForSerialization()`). Since all of *its* dependencies must be either up to date or rebuilt too, I think us pulling out the serialized deps here is redundant regardless. Does that make sense, or am I still missing something?

jrose-apple replied:
I'm surprised that `dependenciesAreUpToDate` adds dependencies to the outer tracker at all. That doesn't seem like its job, and that would explain why this code is here. But having it do that does save us a bit of effort in reading from the cached compiled module twice. It at least deserves to be commented loudly, though, since that name sounds like a predicate.

harlanhaskins replied:
Yeah, `dependenciesAreUpToDate` should probably be renamed here. Good catch noticing that it doesn't check if they actually are up to date 😳

nathawes replied:
I ended up making it not add the dependencies and added them a few levels up in `findOrBuildLoadableModule` and updated it's doc to mention that it adds the dependencies to the dependencyTracker too.

Commit:
[ParseableInterface] Don't report out-of-date dependencies to the parent dependency tracker

We previously added dependencies to the tracker inline while validating a cached
module's dependencies were up to date. If one of its dependencies ended up being
out of date though, we shouldn't have added the previous dependencies, as that
means the dependency list itself was also out of date.

This patch changes the behavior to only add the module's dependencies once we've
verified they're all up to date.

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci please test

jrose-apple commented on include/swift/Serialization/SerializedModuleLoader.h:
```diff
@@ -80,6 +80,12 @@ class SerializedModuleLoaderBase : public ModuleLoader {
 
+  /// Determines if the provided path represents a swiftmodule in either the
+  /// module cache or prebuilt module cache.
+  virtual bool isInModuleCache(StringRef DepPath) {
```
Nitpick (can be follow-up PR): maybe instead of `isInModuleCache` we can just say `isCached`? Or something like `isCachedArtifactForDependencyPurposes` if you want to be explicit.

nathawes replied:
Good point. It should be about how it's used rather than implemented. Updated to `isCached` and changed the doc comment to the prose equivalent of your longer name suggestion.

jrose-apple commented on test/ParseableInterface/ModuleCache/module-cache-init.swift:
```diff
@@ -24,7 +24,7 @@
 // RUN: llvm-bcanalyzer -dump %t/modulecache/LeafModule-*.swiftmodule | %FileCheck %s -check-prefix=CHECK-LEAFMODULE
 // CHECK-LEAFMODULE: {{MODULE_NAME.*blob data = 'LeafModule'}}
-// CHECK-LEAFMODULE: {{FILE_DEPENDENCY.*Swift.swiftmodule([/\\].+[.]swiftmodule)?'}}
+// CHECK-LEAFMODULE-NOT: {{FILE_DEPENDENCY.*Swift.swiftmodule([/\\].+[.]swiftmodule)?'}}
```
:-/ If Swift.swiftmodule started showing up after LeafModule.swiftmodule, we wouldn't catch it this way. I think you'll want something like CHECK-DEPENDS-NEGATIVE below.

nathawes replied:
Fixed!

jrose-apple reviewed (commented):
Looks good to me, only small comments left.

nathawes commented:
@swift-ci please test

nathawes commented:
@swift-ci please smoke test

compnerd commented:
This caused a regression in the Windows build! https://dev.azure.com/compnerd/windows-swift/_build/results?buildId=971

nathawes commented:
Sorry @compnerd! PR to hopefully fix the two parseable interface test failures here: https://github.com/apple/swift/pull/23773
